### PR TITLE
Pattern for Nginx Monitoring

### PR DESCRIPTION
**Pull request by joozero**
*Issue #, if available:* #6

*Description of changes:* Add existing EKS scenario for nginx monitoring 

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: docs/patterns/existing-eks-observability-accelerators/existing-eks-nginx-observability.md [hunk 1]
```diff
@@ -0,0 +1,181 @@
+# Existing EKS Cluster Nginx Observability Accelerator
```
Comment (elamaran11) [inline docs/patterns/existing-eks-observability-accelerators/existing-eks-nginx-observability.md:1]: This documentation is much needed but check the [Java Mon PR](https://github.com/aws-observability/cdk-aws-observability-accelerator/pull/76) for the approach which we need to revamp.

File: lib/existing-eks-nginx-observability-construct/rules.yml [hunk 2]
```diff
@@ -0,0 +1,31 @@
+groups:
```
Comment (elamaran11) [inline lib/existing-eks-nginx-observability-construct/rules.yml:1]: I like this file. I think you should rename it to nginx-rules... and move it to common folder like other PR.

File: lib/existing-eks-nginx-observability-construct/index.ts [hunk 3]
```diff
@@ -0,0 +1,107 @@
+// import { Construct } from 'constructs';
```
Comment (elamaran11) [inline lib/existing-eks-nginx-observability-construct/index.ts:1]: This file is also not required, you might have to add code to existing patterns like single OSS, Existing OSS, Existing Graviton OSS to add a flag `niginx.enabled...` and add rulepaths for AmpAddon which will take care of adding rules. Also make sure you update the opentelemetry.yaml to add nginx `https://github.com/aws-observability/terraform-aws-observability-accelerator/blob/main/modules/eks-monitoring/otel-config/templates/opentelemetrycollector.yaml` config from here.

**Review by elamaran11 — CHANGES_REQUESTED** [review]
@joozero This is AWSome work and this is much needed feature and thankyou for doing this. Apologies if i have not aligned you on the nginx mon support in the issue. I have some strategic feedback to overall align to theme of the accelerator framework :
1. We dont need a new lib, bin files basically not a new pattern, existing OSS, existing Graviton OSS and Single New OSS pattern should have an addendum for Nginx mon support. Thats how we have parity with terraform.
2. We still need a separate doc in both single new folder and existing new folder talking about how to nginx mon to existing pattern.
3. `nginxprometheusaddon.ts` is not needed and flux there should be moved to respective patterns. Rulegroups should be added as paths to AmpAddon and new L3 calls to AMp is not required.
Please reachout for any questions but you can build as similar to Java Mon PR.

**Comment by elamaran11** [comment]
@joozero Please refer to this PR - https://github.com/aws-observability/cdk-aws-observability-accelerator/pull/76 for template to do nginx monitoring.

File: bin/existing-eks-nginx-observability.ts [hunk 4]
```diff
@@ -0,0 +1,8 @@
+import ExistingEksNginxConstruct from '../lib/existing-eks-nginx-observability-construct';
+import { configureApp, errorHandler } from '../lib/common/construct-utils';
+
+const app = configureApp();
+
```
[the remark this answers is not in the document]
Comment (joozero) [reply bin/existing-eks-nginx-observability.ts]: remove this file by referring Java PR

File: lib/existing-eks-nginx-observability-construct/rules.yml (hunk 2 above)
Comment (joozero) [reply lib/existing-eks-nginx-observability-construct/rules.yml]: change name and add to common/resources/amp-config

File: lib/existing-eks-nginx-observability-construct/index.ts (hunk 3 above)
Comment (joozero) [reply lib/existing-eks-nginx-observability-construct/index.ts]: add nginx.enabled function by referring Java PR

File: docs/patterns/existing-eks-observability-accelerators/existing-eks-nginx-observability.md (hunk 1 above)
Comment (joozero) [reply docs/patterns/existing-eks-observability-accelerators/existing-eks-nginx-observability.md:1]: change document based on Java PR

**Comment by elamaran11** [comment]
@joozero Is the PR ready for review. I see not all comments are fixed. Also i see GH actions which needs to be fixed.

File: lib/existing-eks-opensource-observability-pattern/nginxprometheusaddon.ts [hunk 5]
```diff
@@ -0,0 +1,44 @@
+import 'source-map-support/register';
```
Comment (elamaran11) [inline lib/existing-eks-opensource-observability-pattern/nginxprometheusaddon.ts:1]: Is this file still required?

File: lib/existing-eks-opensource-observability-pattern/nginxprometheusaddon.ts [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+import * as blueprints from '@aws-quickstart/eks-blueprints';
```
Comment (elamaran11) [inline lib/existing-eks-opensource-observability-pattern/nginxprometheusaddon.ts:2]: Why do we need this file?

File: lib/existing-eks-opensource-observability-pattern/index.ts [hunk 7]
```diff
@@ -77,10 +78,24 @@ export default class ExistingEksOpenSourceobservabilityPattern {
...
+                manifestPath: __dirname + '/../common/resources/otel-collector-config.yml',
+                manifestParameterMap: {
+                    javaScrapeSampleLimit: 1000,
+                    javaPrometheusMetricsEndpoint: "/metrics"
+                }
+            };
+            ampAddOnProps.ampRules?.ruleFilePaths.push(
+                __dirname + '/../common/resources/amp-config/nginx/alerting-rules.yml'
+            );
+            addOns.push(new NginxPrometheusAddon());
```
Comment (elamaran11) [inline lib/existing-eks-opensource-observability-pattern/index.ts:96]: We should not do this, we should rather use multi flux paths Kustomization. Check the API Server PR for more details.

File: docs/patterns/existing-eks-observability-accelerators/existing-eks-nginx-observability.md [hunk 8]
```diff
@@ -0,0 +1,202 @@
...
+        "GRAFANA_KUBELET_DASH_URL" : "https://raw.githubusercontent.com/aws-observability/aws-observability-accelerator/main/artifacts/grafana-dashboards/eks/infrastructure/kubelet.json",
+        "GRAFANA_NSWRKLDS_DASH_URL" : "https://raw.githubusercontent.com/aws-observability/aws-observability-accelerator/main/artifacts/grafana-dashboards/eks/infrastructure/namespace-workloads.json",
+        "GRAFANA_NODEEXP_DASH_URL" : "https://raw.githubusercontent.com/aws-observability/aws-observability-accelerator/main/artifacts/grafana-dashboards/eks/infrastructure/nodeexporter-nodes.json",
+        "GRAFANA_NODES_DASH_URL" : "https://raw.githubusercontent.com/aws-observability/aws-observability-accelerator/main/artifacts/grafana-dashboards/eks/infrastructure/nodes.json",
+        "GRAFANA_WORKLOADS_DASH_URL" : "https://raw.githubusercontent.com/aws-observability/aws-observability-accelerator/main/artifacts/grafana-dashboards/eks/infrastructure/workloads.json",
+        "GRAFANA_JAVA_JMX_DASH_URL" : "https://raw.githubusercontent.com/aws-observability/aws-observability-accelerator/main/artifacts/grafana-dashboards/eks/java/default.json"
```
Comment (elamaran11) [inline docs/patterns/existing-eks-observability-accelerators/existing-eks-nginx-observability.md:111]: THis doesnt look correct, where is the DASH URL for nginx?

File: docs/patterns/existing-eks-observability-accelerators/existing-eks-nginx-observability.md [hunk 9]
```diff
@@ -0,0 +1,202 @@
...
+```bash
+make build
+make pattern existing-eks-nginx-observability deploy
+```
+
+## Deploy an example application to visualize metrics
+In this section we will deploy sample application and extract metrics using AWS OpenTelemetry collector.
+1. Add the helm incubator repo:
+```
+helm repo add ingress-nginx https://kubernetes.github.io/ingress-nginx
```
Comment (elamaran11) [inline docs/patterns/existing-eks-observability-accelerators/existing-eks-nginx-observability.md:136]: you dont need helm install, You can add nginx addon.

**Review by elamaran11 — CHANGES_REQUESTED** [review]
@joozero Thankyou so much for working on this. The pattern still needs some work 
- Documentation is not accurate
- I still see that you are using a separate flux addon, you dont need that approach, you can use flux multipath. see [this](https://github.com/aws-observability/cdk-aws-observability-accelerator/pull/108) PR 
- Please fix all GH action errors.

**Review by elamaran11 — CHANGES_REQUESTED** [review]
@joozero Adding a comment on `nginx` addon as discussed.

File: lib/existing-eks-opensource-observability-pattern/index.ts (hunk 7 above)
Comment (joozero) [reply lib/existing-eks-opensource-observability-pattern/index.ts]: remove nginxprometheusaddon and simplified

File: lib/existing-eks-opensource-observability-pattern/nginxprometheusaddon.ts (hunk 6 above)
Comment (joozero) [reply lib/existing-eks-opensource-observability-pattern/nginxprometheusaddon.ts]: removed

File: lib/existing-eks-opensource-observability-pattern/nginxprometheusaddon.ts (hunk 5 above)
Comment (joozero) [reply lib/existing-eks-opensource-observability-pattern/nginxprometheusaddon.ts]: removed

File: docs/patterns/single-new-eks-observability-accelerators/single-new-eks-nginx-opensource-observability.md [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+```
+
+2. Enter the following command to create a new namespace:
+```
+kubectl create namespace nginx-ingress-sample
+```
+
+3. Enter the following commands to install NGINX:
+```
+helm install my-nginx ingress-nginx/ingress-nginx \
```
[the remark this answers is not in the document]
Comment (joozero) [reply docs/patterns/single-new-eks-observability-accelerators/single-new-eks-nginx-opensource-observability.md]: @elamaran11 done please take a look

File: docs/patterns/existing-eks-observability-accelerators/existing-eks-nginx-observability.md [hunk 11]
```diff
@@ -0,0 +1,229 @@
...
+```bash
+make build
+make pattern existing-eks-nginx-observability deploy
+```
+
+## Deploy an example Nginx application
+
+In this section we will deploy sample application and extract metrics using AWS OpenTelemetry collector.
+
+1. Add NGINX ingress controller add-on into [lib/single-new-eks-opensource-observability-pattern/index.ts](../../../lib/single-new-eks-opensource-observability-pattern/index.ts) in add-on array.
```
Comment (elamaran11) [inline docs/patterns/existing-eks-observability-accelerators/existing-eks-nginx-observability.md:136]: This is an existing cluster pattern why are we talking about new cluster pattern?

File: docs/patterns/existing-eks-observability-accelerators/existing-eks-nginx-observability.md [hunk 12]
```diff
@@ -0,0 +1,229 @@
...
+                        }
+                    }
+                }
+            }),
+        ];
+```
+
+2. Deploy pattern again 
+```
+make pattern single-new-eks-opensource-observability deploy
```
Comment (elamaran11) [inline docs/patterns/existing-eks-observability-accelerators/existing-eks-nginx-observability.md:171]: this should be existing pattern

File: docs/patterns/existing-eks-observability-accelerators/existing-eks-opensource-observability.md [hunk 13]
```diff
@@ -156,7 +156,8 @@ If you need Java observability you can instead use:
         }
       ]
     },
-    "java.pattern.enabled": true
+    "java.pattern.enabled": true,
+    "nginx.pattern.enabled": true
```
Comment (elamaran11) [inline docs/patterns/existing-eks-observability-accelerators/existing-eks-opensource-observability.md:160]: Remove this because you have separate documentation.

**Review by elamaran11 — CHANGES_REQUESTED** [review]
@joozero AWSome work, couple of minor doc changes, once you fix it, i can run e2e.

File: docs/patterns/existing-eks-observability-accelerators/existing-eks-opensource-observability.md (hunk 13 above)
Comment (joozero) [reply docs/patterns/existing-eks-observability-accelerators/existing-eks-opensource-observability.md:160]: I thought that I can add this because java.pattern.enabled was already existed. Then can I remove both enable code or just nginx?

Comment (elamaran11) [reply docs/patterns/existing-eks-observability-accelerators/existing-eks-opensource-observability.md:160]: just remove the line `"nginx.pattern.enabled": true` 160 from above.

**Review by elamaran11 — CHANGES_REQUESTED** [review]
Couple more

File: docs/patterns/existing-eks-observability-accelerators/existing-eks-nginx-observability.md [hunk 14]
```diff
@@ -0,0 +1,229 @@
...
+    },
+    "nginx.pattern.enabled": true
+  }
+```
+
+8. Once all pre-requisites are set you are ready to deploy the pipeline. Run the following command from the root of this repository to deploy the pipeline stack:
+
+```bash
+make build
+make pattern existing-eks-nginx-observability deploy
```
[the remark this answers is not in the document]
Comment (joozero) [reply docs/patterns/existing-eks-observability-accelerators/existing-eks-nginx-observability.md]: fixed
